packages/application-shell-connectors/src/components/application-state/application-state.js

```diff
@@ -0,0 +1,134 @@
+    // a locale (language + country).
+    locale: user.language,
+    timeZone: user.timeZone || defaultTimeZone,
+  };
+};
+
+// Expose only certain fields as some of them are only meant to
+// be used internally in the AppShell
+const mapProjectToApplicationStateProject = (project, projectDataLocale) => {
+  if (!project) return null;
+
+  return {
+    key: project.key,
+    version: project.version,
+    name: project.name,
+    countries: project.countries,
+    currencies: project.currencies,
+    languages: project.languages,
+    permissions: omit(project.permissions, ['__typename']),
+    // Additional fields that do not belong directly to a project
+    dataLocale: projectDataLocale || null,
+  };
+};
+
+const createApplicationState = defaultMemoize(
+  (environment, user, project, projectDataLocale) => ({
+    environment,
+    user: mapUserToApplicationStateUser(user),
+    project: mapProjectToApplicationStateProject(project, projectDataLocale),
+  })
+);
+
+const ApplicationStateProvider = props => (
+  <Provider
+    value={createApplicationState(
+      props.environment,
+      props.user,
+      props.project,
+      props.projectDataLocale
+    )}
+  >
+    {props.children}
+  </Provider>
+);
+ApplicationStateProvider.displayName = 'ApplicationStateProvider';
+// NOTE: some fields (user, project and projectDataLocale) are optional
+// depending on the render phase of the ApplicationStateProvider.
+// Furthermore, some fields (project, projectDataLocale) might be eventually
+// undefined because in some views (e.g. accounts) we are not in a project context anymore.
+ApplicationStateProvider.propTypes = {
+  // This is the environment configuration coming from `windows.app`
+  environment: PropTypes.shape({
+    frontendHost: PropTypes.string.isRequired,
+    mcApiUrl: PropTypes.string.isRequired,
+    location: PropTypes.string.isRequired,
+    env: PropTypes.string.isRequired,
+    cdnUrl: PropTypes.string.isRequired,
+    servedByProxy: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
+    // ...plus other fields that are specific to each application
+  }).isRequired,
+  user: PropTypes.shape({
+    id: PropTypes.string.isRequired,
+    email: PropTypes.string.isRequired,
+    firstName: PropTypes.string.isRequired,
+    lastName: PropTypes.string.isRequired,
+    language: PropTypes.string.isRequired,
+    timeZone: PropTypes.string,
+    // ...plus other fields that we don't want to expose
+  }),
+  project: PropTypes.shape({
+    key: PropTypes.string.isRequired,
+    version: PropTypes.number,
+    name: PropTypes.string.isRequired,
+    countries: PropTypes.arrayOf(PropTypes.string).isRequired,
+    currencies: PropTypes.arrayOf(PropTypes.string).isRequired,
+    languages: PropTypes.arrayOf(PropTypes.string).isRequired,
+    permissions: PropTypes.object.isRequired,
+    // ...plus other fields that we don't want to expose
+  }),
+  projectDataLocale: PropTypes.string,
```

Comment:
Is this still correct? I thought it was `project.dataLocale`?

Reply:
This is the prop that we pass to the provider. The application state will contain `project.dataLocale` though...I know, it's a bit confusing.

Reply:
I also got confused by this at first but it's okay to keep it around.

What I think we should do is to enable storing the `dataLocale` per project. Right now there is only one `dataLocale` and it is set for all projects - even when the projects might use different languages. So `localStorage` contains `{ dataLocale: 'en' }`.

Instead, I'd store one `dataLocale` for each project in `localStorage` like

`{ dataLocaleByProjectKey: { 'test-with-big-data-44': 'en', 'random-project-2': 'de' } }`.

This is a new feature of course, but it would also help to clean up this code as we'd be passing `dataLocaleByProjectKey` to `ApplicationStateProvider`.

But this would needs some user-research first to determine whether having a per-project data-locale setting is something users actually want.